src/Npgsql.GeoJSON/GeoJsonHandler.cs

```diff
@@ -0,0 +1,796 @@
+        {
+            var builder = new CrsMapBuilder();
+            if ((_option & (GeoJsonOption.ShortCrs | GeoJsonOption.LongCrs)) != GeoJsonOption.None)
+                using (var cmd = new NpgsqlCommand(
+                    "SELECT min(srid), max(srid), auth_name " +
+                    "FROM(SELECT srid, auth_name, srid - rank() OVER(ORDER BY srid) AS range " +
+                    "FROM spatial_ref_sys) AS s GROUP BY range, auth_name ORDER BY 1;", conn))
+                using (var reader = cmd.ExecuteReader())
+                    while (reader.Read())
+                    {
+                        builder.Add(new CrsMapEntry(
+                            reader.GetInt32(0),
+                            reader.GetInt32(1),
+                            reader.GetString(2)));
+                    }
+            return new GeoJsonHandler(_option, builder.Build());
+        }
+    }
+
+    sealed class GeoJsonHandler : NpgsqlTypeHandler<GeoJSONObject>,
+        INpgsqlTypeHandler<Point>, INpgsqlTypeHandler<MultiPoint>,
+        INpgsqlTypeHandler<Polygon>, INpgsqlTypeHandler<MultiPolygon>,
+        INpgsqlTypeHandler<LineString>, INpgsqlTypeHandler<MultiLineString>,
+        INpgsqlTypeHandler<GeometryCollection>,
+        INpgsqlTypeHandler<IGeoJSONObject>,
+        INpgsqlTypeHandler<IGeometryObject>
+    {
+        readonly GeoJsonOption _option;
+        readonly CrsMap _crsMap;
+
+        internal GeoJsonHandler(GeoJsonOption option, CrsMap crsMap)
+        {
+            _option = option;
+            _crsMap = crsMap;
+        }
+
+        GeoJsonOption CrsType => _option & (GeoJsonOption.ShortCrs | GeoJsonOption.LongCrs);
+
+        bool BoundingBox => (_option & GeoJsonOption.BoundingBox) != 0;
+
+        static bool HasSrid(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasSrid) != 0;
+
+        static bool HasZ(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasZ) != 0;
+
+        static bool HasM(EwkbGeometryType type)
+            => (type & EwkbGeometryType.HasM) != 0;
+
+        static bool HasZ(IPosition coordinates)
+            => coordinates.Altitude.HasValue;
+
+        const int SizeOfLength = sizeof(int);
+        const int SizeOfHeader = sizeof(byte) + sizeof(EwkbGeometryType);
+        const int SizeOfHeaderWithLength = SizeOfHeader + SizeOfLength;
+        const int SizeOfPoint2D = 2 * sizeof(double);
+        const int SizeOfPoint3D = 3 * sizeof(double);
+
+        static int SizeOfPoint(bool hasZ)
+            => hasZ ? SizeOfPoint3D : SizeOfPoint2D;
+
+        static int SizeOfPoint(EwkbGeometryType type)
+        {
+            var size = SizeOfPoint2D;
+            if (HasZ(type))
+                size += sizeof(double);
+            if (HasM(type))
+                size += sizeof(double);
+            return size;
+        }
+
+        #region Throw
+
+        static Exception UnknownPostGisType()
```

Comment:
Is there any particular benefit in wrapping the exception creation in static methods, especially for those which are only thrown from a single site? For the ones thrown from multiple site there's at least some value centralizing the string, even the more standard way is to put the strings in a a resx (although that would be overkill...)

Reply:
I thought about something like a throw helper without an additional class, but you're right. So will do. In future I think we should move all strings to resx files as continuation of #685.